iup/__init__.py

```diff
@@ -153,69 +153,63 @@ def to_incident(self, group_cols: tuple[str,] | None) -> IncidentUptakeData:
 
         return IncidentUptakeData(out)
 
+    def insert_rollout(
+        self, rollout: List[dt.date], group_cols: tuple | None
+    ) -> pl.DataFrame:
+        """
+        Insert into cumulative uptake data rows with 0 uptake on rollout dates.
 
-def insert_rollout(
-    frame: pl.DataFrame, rollout: dt.date, group_cols: dict | None
-) -> pl.DataFrame:
-    """
-    Insert into NIS uptake data rows with 0 uptake on the rollout date.
+        Parameters
+        rollout: List[dt.date]
+            list of rollout dates
+        group_cols: tuple[str] | None
+            names of grouping factor columns
 
-    Parameters
-    frame: pl.DataFrame
-        NIS data in the midst of parsing
-    rollout: dt.date
-        rollout date
-    group_cols: dict | None
-        dictionary of the NIS columns for the grouping factors
-        keys are the NIS column names and values are the desired column names
+        Returns
+            cumulative uptake data with rollout rows included
 
-    Returns
-        NIS cumulative data with rollout rows included
+        Details
+        A separate rollout row is added for every grouping factor combination.
+        """
+        frame = self
 
-    Details
-    A separate rollout row is added for every grouping factor combination.
-    """
-    if group_cols is not None:
-        rollout_rows = (
-            frame.select(pl.col(v) for v in group_cols.values())
-            .unique()
-            .with_columns(date=rollout, estimate=0.0)
-        )
-    else:
-        rollout_rows = pl.DataFrame({"date": rollout, "estimate": 0.0})
+        for r in rollout:
+            if group_cols is not None:
+                rollout_rows = (
+                    frame.select(pl.col(v) for v in group_cols)
+                    .unique()
+                    .with_columns(date=r, estimate=0.0)
+                )
+            else:
+                rollout_rows = pl.DataFrame({"date": r, "estimate": 0.0})
 
-    frame = frame.vstack(rollout_rows.select(frame.columns)).sort("date")
+            frame = frame.vstack(rollout_rows.select(frame.columns)).sort("date")
 
-    return frame
+        return frame
 
 
 def extract_group_names(
-    group_cols: List[dict],
-) -> tuple[str,] | None:
+    group_cols: List[List[str]],
```

Comment:
Is there a reason the group columns would be different by dataset?

I find this a little confusing; I would expect you to have a single dataframe, with whatever the stratifying/grouping variables are (season, geography, etc.) and then you pass that to the model

But this makes it seem like different data sets are in different data frames, with potentially different schema?

Reply:
Different data sets are in different data frames, at least at first. E.g. if you want to train a model on covid vaccine uptake from 2022 - 2024, this spans two data sets (udsf-9v7b and ksfb-ug5d, specifically) that come out of the NIS API separately. In `main.py`, `split_train_test` is the function that concatenates them into the same frame. Before `split_train_test` is called, the data sets are in separate dataframes. So the desired grouping factors are specified for each dataframe separately (and `extract_group_names` makes sure that they match). 

Maybe a better way to do this is to have one standalone entry for desired grouping factors across all data sets in the config.yaml, rather than one under each data set? This way the user doesn't need to type out the same grouping factors repeatedly for each data set, since they must match across all data sets anyway.

Reply:
`nisapi` will deliver you the stuff in udsf-9v7b and ksfb-ug5d as rows in a single dataframe, which I think means you can count on all datasets having the same schema, grouping columns, etc.

Reply:
I think this has been addressed in the latest updates to this PR - the user specifies which columns they want to keep (including which grouping factors they want to use), but the interface for doing so in the config is easier because we are now assuming all data sets have the same schema. `extract_group_names` is gone, actually.